### DISCUSSION TITLE
Subprotocols

### DISCUSSION THREAD

**Discussion by davidbrochart** [frankie567/httpx-ws#23] [Q&A]
`wsproto` seems to [support subprotocols](https://python-hyper.org/projects/wsproto/en/stable/api.html?highlight=subprotocols#wsproto.events.Request), but I'm not sure where to pass them in `httpx-ws`. Or maybe it's not supported yet?

**Comment by frankie567** [comment]
Well, there is no dedicated option for that. Actually, subprotocols seem to be declared in a `Sec-WebSocket-Protocol` protocol header when making the first request. So what you could is to manually set this header:

```py
from httpx_ws import aconnect_ws

async with aconnect_ws("http://localhost:8000/ws", headers={"Sec-WebSocket-Protocol": "sub1, sub2"}) as ws:
    message = await ws.receive_text()
    print(message)
    await ws.send_text("Hello!")
```

**Reply by davidbrochart** [reply]
Thanks, that seems to work.
The server is supposed to reply with the selected subprotocol from the list of subprotocols sent by the client. Do you know how to get it?

**Reply by frankie567** [reply]
Hmm, in the current state, I don't think that's possible. I suppose the client should emit [AcceptConnection](https://python-hyper.org/projects/wsproto/en/stable/api.html#wsproto.events.AcceptConnection) when opened so we can get that info 🤔

**Reply by davidbrochart** [reply]
Yes it looks like it, since this event has the accepted subprotocol.
It would be a nice addition to `httpx-ws` in general, and for me in particular :smile:

**Reply by frankie567** [reply]
Was thinking about this... In terms of API, it would a bit annoying/cumbersome in most cases to have to first handle an `AcceptConnection` event.

What would you think of this: when opening the connection, we would set the `accept` event as attribute of `WebSocketSession`. You can read it if you need, but if you don't care, you don't have to worry about it.

```py
from httpx_ws import aconnect_ws

async with aconnect_ws("http://localhost:8000/ws", subprotocols=["sub1", "sub2"]) as ws:
    print(f"Subprotocol {ws.accept.subprotocol}")
    message = await ws.receive_text()
    print(message)
    await ws.send_text("Hello!")
```

Or maybe just directly `subprotocol`, maybe we don't need the event 🤔 

```py
from httpx_ws import aconnect_ws

async with aconnect_ws("http://localhost:8000/ws", subprotocols=["sub1", "sub2"]) as ws:
    print(f"Subprotocol {ws.subprotocol}")
    message = await ws.receive_text()
    print(message)
    await ws.send_text("Hello!")
```

**Reply by davidbrochart** [reply]
I think `ws.subprotocol` (without `.accept`) is shorter and better.
